### PR TITLE
Cover regrets/weights new participants edge cases

### PR DESCRIPTION
**Pull request by guilherme-brandao**
Consider the edge cases:
- If all inferers are new, ignore the forecasters, and set the weights of all inferers to 1, irrespectively of the number of not-new forecasters.
- If one inferer is not new and all other inferers are new, set the weight of the not-new inferer to 1, and all others (including the forecasters, because we don’t need them anyway in this case) get a zero weight.
- If >1 inferer is not new, calculate weight as in the paper for all not-new workers (inferer and forecaster), and don’t include any new workers in the regret calculation.

### CODE REVIEW

File: x/emissions/keeper/inference_synthesis/synth_palette_weight.go [hunk 1]
```diff
@@ -99,31 +111,48 @@ func (p *SynthPalette) CalcWeightedInference(weights RegretInformedWeights) (Inf
...
-		runningUnnormalizedI_i, sumWeights, err = AccumulateWeights(
-			p.InferenceByWorker[inferer],
-			weights.inferers[inferer],
-			p.InfererRegrets[inferer].noPriorRegret,
-			p.AllInferersAreNew,
-			runningUnnormalizedI_i,
-			sumWeights,
-		)
+	// If there is only one not-new inferer, then just consider that inferer
+	if p.SingleInfererNotNew != "" {
```
Comment (kpeluso) [inline x/emissions/keeper/inference_synthesis/synth_palette_weight.go:115]: it's not wrong this way but I would prefer to use the booleans as booleans as opposed to derivative properties such as this (this is only populated if those booleans are of a certain value)

File: x/emissions/keeper/inference_synthesis/synth_palette_weight.go [hunk 2]
```diff
@@ -64,23 +65,34 @@ func (p *SynthPalette) CalcWeightsGivenWorkers() (RegretInformedWeights, error)
...
 
 	forecasterWeights := make(map[Worker]Weight)
-	for _, worker := range p.Forecasters {
-		forecasterWeight, err := CalcWeightFromNormalizedRegret(normalizedForecasterRegrets[worker], maxRegret, p.PNorm, p.CNorm)
-		if err != nil {
-			return RegretInformedWeights{}, errorsmod.Wrapf(err, "Error calculating forecaster weight")
+	if len(p.ForecasterRegrets) > 0 {
+		for _, worker := range p.Forecasters {
+			var forecasterWeight = alloraMath.ZeroDec()
+			if moreThanOneNotNewInferer && !p.ForecasterRegrets[worker].noPriorRegret {
```
Comment (kpeluso) [inline x/emissions/keeper/inference_synthesis/synth_palette_weight.go:88]: for the check of `moreThanOneNotNewInferer` here and above, seems like you can simply skip both loops entirely if it's false, then whenever you check the weight you just check for inclusion before you do calculations with the weight. this fits with the pattern we established of using `palette.[inferers | forecasters]` to guide computation in all steps

File: x/emissions/keeper/inference_synthesis/types.go [hunk 3]
```diff
@@ -68,13 +68,12 @@ type SynthPalette struct {
...
-	AllForecastersAreNew bool
-	AllWorkersAreNew     bool // Simple conjunction of the two above
-	NetworkCombinedLoss  Loss
-	Epsilon              alloraMath.Dec
-	FTolerance           alloraMath.Dec
-	PNorm                alloraMath.Dec
-	CNorm                alloraMath.Dec
+	ForecasterRegrets   map[Worker]*StatefulRegret
+	AllInferersAreNew   bool
+	SingleInfererNotNew Worker
```
Comment (kpeluso) [inline x/emissions/keeper/inference_synthesis/types.go:73]: i get that it's more efficient this way to cache this value. it does feel cleaner to check this in the accumulator using the existing "noPriorRegret" prop. I can go either way

File: x/emissions/keeper/inference_synthesis/synth_palette_weight.go [hunk 4]
```diff
@@ -64,23 +65,34 @@ func (p *SynthPalette) CalcWeightsGivenWorkers() (RegretInformedWeights, error)
 		}
 	}
 
+	moreThanOneNotNewInferer := p.SingleInfererNotNew == ""
```
Comment (kpeluso) [inline x/emissions/keeper/inference_synthesis/synth_palette_weight.go:68]: you use this condition check in multiple places. You should probs have a palette method for it

**Review by kpeluso — APPROVED** [review]
approving because it accurately implements the intentions. my previous comments can just possibly make the code better

File: x/emissions/keeper/inference_synthesis/synth_palette_weight.go (hunk 1 above)
Comment (guilherme-brandao) [reply x/emissions/keeper/inference_synthesis/synth_palette_weight.go:115]: Yeah, also thought about this but was opposed to adding a new property for the palette just for that... maybe it's a better idea to add...

**Comment by br4e** [comment]
Proposed changes to address kenny's comments and also my own concerns: 

see here https://github.com/allora-network/allora-chain/pull/338/files

**Pull request merged**